crates/astria-sequencer/src/ibc/ics20_withdrawal.rs

```diff
@@ -51,6 +53,49 @@ fn withdrawal_to_unchecked_ibc_packet(
     )
 }
 
+async fn ics20_withdrawal_check_stateful_bridge_account<S: StateReadExt + 'static>(
+    action: &action::Ics20Withdrawal,
+    state: &S,
+    from: Address,
+) -> Result<()> {
+    // bridge address checks:
+    // - if the sender of this transaction is not a bridge account, and the tx `bridge_address`
+    //   field is None, don't need to do any bridge related checks as it's a normal user withdrawal.
+    // - if the sender of this transaction is a bridge account, and the tx `bridge_address` field is
+    //   None, check that the withdrawer address is the same as the transaction sender.
+    // - if the tx `bridge_address` field is Some, check that the `bridge_address` is a valid
+    //   bridge, and check that the withdrawer address is the same as the transaction sender.
+
+    let is_sender_bridge = state
+        .get_bridge_account_rollup_id(&from)
+        .await
+        .context("failed to get bridge account rollup id")?
+        .is_some();
+
+    if !is_sender_bridge && action.bridge_address.is_none() {
+        return Ok(());
+    }
+
+    // if `action.bridge_address` is Some, but it's not a valid bridge account,
+    // the `get_bridge_account_withdrawer_address` step will fail.
+    let bridge_address = action.bridge_address.unwrap_or(from);
+
+    let Some(withdrawer) = state
+        .get_bridge_account_withdrawer_address(&bridge_address)
+        .await
+        .context("failed to get bridge withdrawer")?
+    else {
+        bail!("bridge address must have a withdrawer address set");
+    };
```

Comment:
Would this cause sync failures with old withdrawals? If the withdrawer isn't set should we just assume it's the same as bridge address?

Reply:
bridge accounts initialized before these changes would not have sudo/withdrawer set, so they would have to submit a `BridgeSudoChange` action to set them before this would work. i originally implemented it like you said but i didn't like it as it's better to force the addresses to be set